dotCMS/src/main/java/com/dotcms/concurrent/scheduler/DotScheduler.java

```diff
@@ -0,0 +1,129 @@
+     * @param recurringTasks {@link DotRecurringTask} array
+     * @return DotScheduler
+     */
+    DotScheduler registerStatefulRecurringTask  (final DotRecurringTask... recurringTasks);
+
+    /**
+     * Register a recurring task for stateless (recurring execution on each node in a server),
+     * it will be executed based on the cron expression on the {@link DotRecurringTask}
+     * @param recurringTasks
+     * @return DotScheduler
+     */
+    DotScheduler registerStatelessRecurringTask (final DotRecurringTask... recurringTasks);
+
+    /**
+     * Unregister a task
+     * @param instanceId {@link String} identifier of the task instance
+     * @return DotTask null if does not exists
+     */
+    DotTask       unregisterTask                 (final String instanceId);
+
+    /**
+     * Starts the scheduler, if already start won't do anything.
+     * Note: it will be shutdown when the server shutdowns
+     */
+    void        start   ();
+
+    /**
+     * Restart the scheduler, it is useful when a new task is registered or unregistered once the scheduler already started
+     */
+    void        restart ();
+
+    void executeOnce(final TaskInstance<DotTask> taskInstance, final ExecutionContext executionContext);
+
+    /**
+     * Fires Task, if the task does not exists an NotTaskFoundException will be thrown
+     * @param dotTask {@link DotTask} dot task
+     */
+    void        fire(final DotTask dotTask);
+
+    /**
+     * Fires Task, if the task does not exists an NotTaskFoundException will be throw
+     * @param dotTask {@link DotTask} dot task
+     * @param delay {@link Duration} delay to wait for the task execution
+     */
+    void        fire(final DotTask dotTask, final Duration delay);
+
+    // todo: check if when re-fire is needed with a delay, instead of the sending the instanceId, we need the actual instance (the DotTask)
+
+    /**
+     * Cancels a task,  if the task does not exists an NotTaskFoundException will be throw
+     * @param instanceId {@link String} identifier of the task instance
+     */
+    void        cancel(final String instanceId);
+
+    /**
+     * Returns the status of scheduler
+     * @return SchedulerStatus
+     */
+    SchedulerStatus status();
+
+    /**
+     * Returns the status of a task
+     * @param instanceId {@link String} identifier of the task instance
+     * @return TaskStatus
+     */
+    TaskStatus taskStatus(final String instanceId);
+
+    /**
+     * Status for the Scheduler
+     */
+    enum SchedulerStatus {
+        STOPPED, RUNNING
+    }
+
+    /**
+     * Status for a task
+     */
+    enum TaskStatus {
+        STOPPED, RUNNING, FAILED, CANCELED, UNKNOWN
+    }
+
+    /**
+     * Returned when the task requested does not exists
+     */
+    class NotTaskFoundException extends RuntimeException {
```

Comment:
Codacy found an issue: [Classes implementing Serializable should set a serialVersionUID](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=6006494)